cadc-pkg-server/src/main/java/org/opencadc/pkg/server/PackageWriter.java

```diff
@@ -131,110 +133,102 @@ public void write(PackageItem packageItem) throws IOException, InterruptedExcept
-     * @return
-     * @throws IOException
+     * Write a directory to the archive.
      */
-    private void setArchiveEntryHeader(String relativePath, long contentLength, Date lastModified)
-        throws IOException {
-
-        // create entry (metadata) to be put to archive stream
-        log.debug("next package entry: " + relativePath + "," + contentLength + "," + lastModified);
-        ArchiveEntry e = createEntry(relativePath, contentLength, lastModified);
+    private void writeDirectory(PackageItem packageItem)
+            throws IOException {
+        log.debug("write directory: " + packageItem.getRelativePath());
 
-        // put archive entry to stream
-        aout.putArchiveEntry(e);
+        ArchiveEntry archiveEntry = createDirectoryEntry(packageItem.getRelativePath());
+        archiveOutputStream.putArchiveEntry(archiveEntry);
+        archiveOutputStream.closeArchiveEntry();
     }
 
     /**
-     * Write the content of a file:// URL to the current ArchiveOutputStream
-     * @param fileURL - URL of file item being written to
-     * @param relativePath - relative path of file inside package beingw written.
-     * @throws IOException
+     * Write a file to the archive with the file content is a local file.
      */
-    private void writeFileContentEntry(URL fileURL, String relativePath) throws IOException {
-        Path filePath = FileSystems.getDefault().getPath(fileURL.getPath());
-        log.debug("filePath from fileURL: " + filePath.toString() + ": " + fileURL.toString());
+    private void writeFile(PackageItem packageItem)
+            throws IOException {
+        log.debug(String.format("write file %s to %s", packageItem.getContent(), packageItem.getRelativePath()));
 
+        URL fileURL = packageItem.getContent();
+        Path filePath = FileSystems.getDefault().getPath(fileURL.getPath());
         long contentLength = Files.size(filePath);
         FileTime lastMod = Files.getLastModifiedTime(filePath);
 
         Date lastModified = new Date(lastMod.toMillis());
-        boolean openEntry = false;
-
-        try {
-            // This gets set to true unless setArchiveEntryHeader throws an error
-            setArchiveEntryHeader(relativePath, contentLength, lastModified);
+        String relativePath = packageItem.getRelativePath();
+        ArchiveEntry archiveEntry = createFileEntry(relativePath, contentLength, lastModified);
 
-            openEntry = true;
-
-            // copy the file to archive output stream
-            Files.copy(filePath, aout);
-        } finally {
-            if (openEntry) {
-                // header was written but body was not, try to close gracefully
-                aout.closeArchiveEntry();
-            }
-        }
+        // put archive entry to stream
+        archiveOutputStream.putArchiveEntry(archiveEntry);
 
+        // copy the file to archive output stream
+        Files.copy(filePath, archiveOutputStream);
```

Comment:
I'm pretty sure MultiBufferIO.copy is faster than this builtin function because it reads and writes in separate threads. It can make a difference when one of the streams (usually the client stream) is slower than the other.

You can just use `fileURL.openStream()` in this case.

Reply:
MultiBufferIO.copy is already used for HTTP urls, updated to use it for file urls as well.